### PR TITLE
Add ability to override/merge items in local .DEREK.yml

### PR DESCRIPTION
**Pull request by Waterdrips**
## Description

We can now provide some local overrides from our Local DEREK.yml,
these overrides are then merged with the remote DEREK.yaml.
This allows us to add specific users to specific repos, or add
features for specific ones too. Any slices are merged and the
single values are overridden by the local values.

Signed-off-by: Alistair Hey <alistair@heyal.co.uk>

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md))
#94 

## How Has This Been Tested?

Unit tests written, and this was tested on github using this repo: https://github.com/Waterdrips/heyal-ofc-customers

I setup a .DEREK.yml that had me as a contributor and a load of general config, but no "features"
pointed the Redirect at .DEREK-OVERRIDE.yml in the same repo, which added @alexellis as a contributor, plus it had the features listed.

Then I created a non signed commit and PR, DCO check was done (features in redirect file)
I was able to use the derek commands (local file)

And Alex was able to use Derek commands (set in redirect file)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: handler/permissions_handler.go [hunk 1]
```diff
@@ -55,12 +55,12 @@ func PermittedUserFeature(attemptedFeature string, config *types.DerekRepoConfig
...
 	req, err := http.NewRequest(http.MethodGet, url, nil)
 	if err != nil {
-		log.Fatalln(err)
+		log.Fatalln(fmt.Sprintf("unable to make request to %q, %e", url, err))
 	}
 
 	res, resErr := client.Do(req)
 	if resErr != nil {
-		log.Fatalln(resErr)
+		log.Fatalln(fmt.Sprintf("could not action request url:%q, err:%s", url, resErr.Error()))
```
Comment (alexellis) [inline handler/permissions_handler.go]: Add spacing between `url:%` for both parts please

File: handler/permissions_handler.go [hunk 2]
```diff
@@ -128,27 +128,35 @@ func GetRepoConfig(owner string, repository string) (*types.DerekRepoConfig, err
 }
 
 func buildDerekConfig(client http.Client, bytesConfig []byte) (*types.DerekRepoConfig, error) {
```
Comment (alexellis) [inline handler/permissions_handler.go]: How many redirects do we allow?

Comment (alexellis) [reply handler/permissions_handler.go]: How are we preventing cyclic redirects?

File: types/merge_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		t.Errorf("Got error for a single plan, expected no error: %s", err.Error())
+		t.Fail()
+	}
+
+	if config2.ContributingURL != configOut.ContributingURL {
+		t.Errorf("ContributingURL want: %s, but got: %s", config2.ContributingURL, configOut.ContributingURL)
+	}
+}
+
+func Test_mergeDerekRepoConfigs_MergeEmptyItemsFromBoth(t *testing.T) {
```
Comment (alexellis) [inline types/merge_test.go]: What happens if the 2nd repo has no ContributingURL ? Can you show that the base ContributingURL is still honoured?

Also check all structs for `omitempty` please

**Comment by Waterdrips** [comment]
> Looks good so far. I think we need to add an abuse-prevention technique for cyclic redirects or redirects which are too deep.

Currently it only follows the local file's redirect url (only goes 1 layer deep)
We could add the ability to follow redirects, adding a max depth?

File: handler/permissions_handler.go (hunk 2 above)
Comment (Waterdrips) [reply handler/permissions_handler.go]: It only follows the first url's redirect.

File: types/merge_test.go (hunk 3 above)
Comment (Waterdrips) [reply types/merge_test.go]: Is that not what the config1.Redirect is showing? It doesnt exist in the second so it's not overridden.

File: types/merge.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package types
+
+import (
+	"github.com/imdario/mergo"
+)
+
+func MergeDerekRepoConfigs(localConfig, remoteConfig DerekRepoConfig) (DerekRepoConfig, error) {
+
+	mergeErr := mergo.Merge(&remoteConfig, &localConfig, mergo.WithAppendSlice)
```
Comment (alexellis) [inline types/merge.go]: Are we going to have the same issues we have with ofc-bootstrap? I saw some tests, do we need more?

**Comment by alexellis** [comment]
I suspect the messages list might break @Waterdrips, can you add some tests for union/intersection?

**Comment by Waterdrips** [comment]
Good spot. Iv just written a failing test.

I'll get a pr in to sort that messages list soon

**Pull request merged**